### PR TITLE
RA-1638: Added new (Bahmni) Appointments dashboard widget capability.

### PR DESCRIPTION
**Pull request by mozzy11**
Ticket: https://issues.openmrs.org/browse/RA-1638

### CODE REVIEW

File: omod/src/main/web/dashboardwidgets/appointments/appointments.controller.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            this.maxResults =  this.maxResults + 1;
+            let appoints = {startDatetime: appointmnet.startDateTime,
+                           startTime :this.convertTime(appointmnet.service.startTime),
+                           endTime :this.convertTime(appointmnet.service.endTime),
+                           ServiceType :appointmnet.serviceType  
+            };
+            this.appointments.push(appoints);
+
+            if(this.maxResults == this.getMaxRecords()){
+                cont = false;
```
[the remark this answers is not in the document]
Comment (mozzy11) [reply omod/src/main/web/dashboardwidgets/appointments/appointments.controller.js]: yes , i actually made that comment on the ticket, if we can just pass that param in the request body , the better

File: omod/src/main/web/dashboardwidgets/appointments/appointments.controller.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        this.appointments = [];
+        this.serverUrl = "";
+        this.startDate = new Date();
+        this.openmrsRest.setBaseAppPath("/coreapps");
+        this.openmrsRest.getServerUrl().then((result) => {
+            this.serverUrl = result;
+        });
+
+        this.openmrsRest.create('appointments/search', {
+            patient: this.config.patientUuid,
```
[the remark this answers is not in the document]
Comment (mozzy11) [reply omod/src/main/web/dashboardwidgets/appointments/appointments.controller.js]: sure ,

File: omod/src/main/web/dashboardwidgets/appointments/appointments.controller.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (sukreet) [reply omod/src/main/web/dashboardwidgets/appointments/appointments.controller.js]: The parameter is `limit`

File: omod/src/main/web/dashboardwidgets/appointments/appointments.controller.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        if (appointmnets.length > 0){
+        let cont = true;
+        angular.forEach(appointmnets, (appointmnet) => {
+            if(cont){
+            this.maxResults =  this.maxResults + 1;
+            let appoints = {startDatetime: appointmnet.startDateTime,
+                           startTime :this.convertTime(appointmnet.service.startTime),
+                           endTime :this.convertTime(appointmnet.service.endTime),
+                           ServiceType :appointmnet.serviceType  
+            };
```
[the remark this answers is not in the document]
Comment (mozzy11) [reply omod/src/main/web/dashboardwidgets/appointments/appointments.controller.js]: i think i had not seen the time property on appointment.startDateTime. ill correct it

File: omod/src/main/web/dashboardwidgets/appointments/appointments.html [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<ul>
+    <li ng-if="$ctrl.appointments.length != 0" ng-repeat="appointment in $ctrl.appointments track by $index">
+      <div >{{appointment.date| date:'mediumDate'}}   {{appointment.startTime | date:'h:mm a'}} - {{appointment.endTime | date:'h:mm a'}} {{appointment.ServiceType  | translate }}</div>
```
Comment (sukreet) [inline omod/src/main/web/dashboardwidgets/appointments/appointments.html]: For consistency can we please change the format slightly.  
`      <div >{{appointment.date| date:'dd.MMM.yyyy'}},  {{appointment.startTime | date:'h:mm a'}} - {{appointment.endTime | date:'h:mm a'}}, {{appointment.ServiceType  | translate }}</div>
`

File: omod/src/main/resources/apps/dashboardWidgets_app.json [hunk 5]
```diff
@@ -120,5 +120,31 @@
...
+    "config": {
+      "widget": "appointments",
+      "icon": "icon-calendar",
+      "label": "APPOINTMENTS",
+      "editPrivilege": "App: appointmentschedulingui.viewAppointments",
+      "maxRecords": "5",
+      "editIcon": "icon-share-alt" ,
+      "detailsUrl" : "appointments/appointments"
+           },
+    "extensions": [
```
[the remark this answers is not in the document]
Comment (mozzy11) [reply omod/src/main/resources/apps/dashboardWidgets_app.json]: sure , i removed it

File: omod/src/main/web/dashboardwidgets/appointments/appointments.html (hunk 4 above)
Comment (mozzy11) [reply omod/src/main/web/dashboardwidgets/appointments/appointments.html]: Alright

File: omod/src/main/web/dashboardwidgets/appointments/appointments.controller.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    // Number of maximum records 
+    getMaxRecords() {
+        if (this.config.maxRecords == '' || angular.isUndefined(this.config.maxRecords)) {
+            return 5;
+        } else {
+            return this.config.maxRecords;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (mozzy11) [reply omod/src/main/web/dashboardwidgets/appointments/appointments.controller.js]: cool

[the remark this answers is not in the document]
Comment (sukreet) [reply omod/src/main/web/dashboardwidgets/appointments/appointments.controller.js]: Thanks

File: omod/src/main/web/dashboardwidgets/appointments/appointments.html (hunk 4 above)
Comment (sukreet) [reply omod/src/main/web/dashboardwidgets/appointments/appointments.html]: Thanks

Comment (mozzy11) [reply omod/src/main/web/dashboardwidgets/appointments/appointments.html]: done

File: omod/src/main/web/dashboardwidgets/appointments/appointments.controller.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (mozzy11) [reply omod/src/main/web/dashboardwidgets/appointments/appointments.controller.js]: done

**Comment by sukreet** [comment]
@mks-d  Can you please have a look at this PR.

File: omod/src/main/resources/apps/dashboardWidgets_app.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (mks-d) [reply omod/src/main/resources/apps/dashboardWidgets_app.json]: @mozzy11 however it is useful to have a sample configuration "somewhere". Could you perhaps document this on the wiki where all other dashboard widgets are documented?

[the remark this answers is not in the document]
Comment (mozzy11) [reply omod/src/main/resources/apps/dashboardWidgets_app.json]: yes @mks-d . We were actually still debatng with @sukreet where exactly to have this documented ,whether in openmrs or bamni. Thanks Am going to document it.

**Comment by mozzy11** [comment]
@mks-d  , i have added it to the documentation 
https://wiki.openmrs.org/display/docs/Patient+Summary+Widget+Documentation#PatientSummaryWidgetDocumentation-BamhniAppointmentsWidget.

**Pull request merged**